File: sandbox/broker/internal/handler/handler.go
// Package handler bridges the wire layer, policy set, and credential
// backend. It does NOT bind to the socket — that belongs to the server
// package, so policy logic stays unit-testable without spinning up UDS.
// Parity reference: src/claude_config/egress_broker/handler.py.
package handler

import (
	"bytes"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/AudensAurantius/claude-config/sandbox/broker/internal/credentials"
	"github.com/AudensAurantius/claude-config/sandbox/broker/internal/policy"
	"github.com/AudensAurantius/claude-config/sandbox/broker/internal/wire"
)

// Handler enforces policy and forwards to the upstream.
type Handler struct {
	Policies   *policy.PolicySet
	Backend    credentials.Backend
	HTTPClient *http.Client // nil = construct per-request with policy timeout
	Logger     *slog.Logger
}

// New returns a Handler with a default logger if Logger is nil.
func New(policies *policy.PolicySet, backend credentials.Backend) *Handler {
	return &Handler{Policies: policies, Backend: backend, Logger: slog.Default()}
}

// Handle processes one request and returns a wire payload.
func (h *Handler) Handle(req *wire.Request) map[string]any {
	pol := h.Policies.Get(req.Alias)
	if pol == nil {
		return wire.EncodeError("policy-denied", fmt.Sprintf("unknown alias: %q", req.Alias))
	}
	if denial := checkPolicy(req, pol); denial != "" {
		return wire.EncodeError("policy-denied", denial)
	}
	secret, err := h.Backend.Fetch(pol.Credential.Path)
	if err != nil {
		h.Logger.Error("credential fetch failed", "alias", req.Alias, "err", err)
		return wire.EncodeError("upstream-failed", "credential backend failure")
	}
	outURL, outHeaders, outBody := buildOutbound(req, pol, secret)
	resp, err := h.doUpstream(pol, req.Method, outURL, outHeaders, outBody)
	if err != nil {
		h.Logger.Error("upstream call failed", "alias", req.Alias, "err", err)
		return wire.EncodeError("upstream-failed", err.Error())
	}
	return wire.EncodeResponse(resp)
}

func checkPolicy(req *wire.Request, pol *policy.Policy) string {
	if _, ok := pol.Constraints.Methods[req.Method]; !ok {
		return fmt.Sprintf("method %q not in allowed methods", req.Method)
	}
	pathOnly := req.Path
	if i := strings.IndexByte(pathOnly, '?'); i >= 0 {
		pathOnly = pathOnly[:i]
	}
	if !pol.Constraints.PathAllowed(pathOnly) {
		return fmt.Sprintf("path %q not in allowed paths", pathOnly)
	}
	if len(req.Body) > pol.Constraints.MaxRequestBytes {
		return fmt.Sprintf("request body %d bytes exceeds max_request_bytes %d",
			len(req.Body), pol.Constraints.MaxRequestBytes)
	}
	return ""
}

// buildOutbound returns the full upstream URL, header map, and body to send.
func buildOutbound(req *wire.Request, pol *policy.Policy, secret string) (string, http.Header, []byte) {
	stripped := pol.Constraints.StrippedHeaders(pol.Credential.Attach)
	out := make(http.Header, len(req.Headers)+2)
	for name, values := range req.Headers {
		if _, drop := stripped[strings.ToLower(name)]; drop {
			continue
		}
		// Preserve multi-value semantics: net/http represents repeats as a slice.
		for _, v := range values {
			out.Add(name, v)
		}
	}
	out.Set("Host", pol.Upstream.Host)

	attach := pol.Credential.Attach
	path := req.Path
	switch attach.Type {
	case "header":
		out.Set(attach.Name, secret)
	case "bearer":
		out.Set("Authorization", "Bearer "+secret)
	case "query":
		path = appendQuery(path, attach.Name, secret)
	}
	upstreamURL := fmt.Sprintf("%s://%s:%d%s", pol.Upstream.Scheme, pol.Upstream.Host, pol.Upstream.Port, path)
	return upstreamURL, out, req.Body
}

func appendQuery(path, key, value string) string {
	u, err := url.Parse(path)
	if err != nil {
		// Should never happen — wire already validated leading '/'.
		return path
	}
	q := u.Query()
	q.Add(key, value)
	u.RawQuery = q.Encode()
	return u.String()
}

func (h *Handler) doUpstream(pol *policy.Policy, method, fullURL string, headers http.Header, body []byte) (*wire.Response, error) {
	var bodyReader io.Reader
	if len(body) > 0 {
		bodyReader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, fullURL, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}
	req.Header = headers
	// net/http sets Host from URL.Host by default; honor an explicit Host header.
	if h := headers.Get("Host"); h != "" {
		req.Host = h
	}

	client := h.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: time.Duration(pol.Constraints.TimeoutSeconds) * time.Second}
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read upstream body: %w", err)
	}
	// Group multi-value response headers faithfully.
	grouped := make(map[string][]string, len(resp.Header))
	for k, vs := range resp.Header {
		grouped[k] = append([]string(nil), vs...)
	}
	return &wire.Response{Status: resp.StatusCode, Headers: grouped, Body: respBody}, nil
}

File: src/claude_config/egress_broker/handler.py
"""Request handler — policy enforcement and upstream HTTPS forwarding.

The handler is the integration point between the wire protocol, the
policy set, and the credential backend. It does NOT bind to the socket
(see ``server.py``) — that decoupling keeps the policy logic unit-testable
without spinning up a UDS.
"""

from __future__ import annotations

import http.client
import logging
import ssl
import urllib.parse
from dataclasses import dataclass
from typing import Any

from .credentials import CredentialBackend, CredentialError
from .policy import Policy, PolicySet
from .wire import Request, Response

logger = logging.getLogger(__name__)


@dataclass
class Handler:
    """Stateful handler bundling policy + credential backend."""

    policies: PolicySet
    credential_backend: CredentialBackend
    ssl_context: ssl.SSLContext | None = None

    def __post_init__(self) -> None:
        """Build an SSLContext if the caller didn't supply one."""
        if self.ssl_context is None:
            self.ssl_context = ssl.create_default_context()

    def handle(self, req: Request) -> dict[str, Any]:
        """Process a single sandbox request and return a wire response JSON.

        Returns a JSON object suitable for ``encode_frame``. The returned
        object is either a successful ``Response.to_json()`` or an error
        envelope ``{"error": <code>, "message": <detail>}``.
        """
        policy = self.policies.get(req.alias)
        if policy is None:
            return _err("policy-denied", f"unknown alias: {req.alias!r}")

        denial = _check_policy(req, policy)
        if denial is not None:
            return _err("policy-denied", denial)

        try:
            secret = self.credential_backend.fetch(policy.credential.path)
        except CredentialError as e:
            logger.error("credential fetch failed for alias=%s: %s", req.alias, e)
            return _err("upstream-failed", "credential backend failure")

        headers, path = _build_outbound(req, policy, secret)

        try:
            resp = _do_upstream(
                policy=policy,
                method=req.method,
                path=path,
                headers=headers,
                body=req.body,
                ssl_context=self.ssl_context,
            )
        except (OSError, http.client.HTTPException, ssl.SSLError) as e:
            logger.error("upstream call failed for alias=%s: %s", req.alias, e)
            return _err("upstream-failed", f"{type(e).__name__}: {e}")

        return resp.to_json()


def _check_policy(req: Request, policy: Policy) -> str | None:
    """Return a denial reason or None if the request is allowed."""
    if req.method not in policy.constraints.methods:
        return f"method {req.method!r} not in allowed methods {sorted(policy.constraints.methods)}"
    # path may include a querystring; check the path portion only against globs
    parsed = urllib.parse.urlsplit(req.path)
    if not policy.constraints.path_allowed(parsed.path):
        return f"path {parsed.path!r} not in allowed paths {list(policy.constraints.paths)}"
    if len(req.body) > policy.constraints.max_request_bytes:
        return (
            f"request body {len(req.body)} bytes exceeds max_request_bytes "
            f"{policy.constraints.max_request_bytes}"
        )
    return None


def _build_outbound(req: Request, policy: Policy, secret: str) -> tuple[dict[str, str], str]:
    """Construct the outbound header dict and path with credential attached."""
    stripped = policy.constraints.stripped_headers(policy.credential.attach)
    out_headers: dict[str, str] = {}
    for name, values in req.headers.items():
        if name.lower() in stripped:
            continue
        # http.client accepts a flat dict; collapse multi-value with comma per RFC 7230.
        out_headers[name] = ", ".join(values)
    out_headers["Host"] = policy.upstream.host

    attach = policy.credential.attach
    path = req.path
    if attach.type == "header":
        out_headers[attach.name] = secret
    elif attach.type == "bearer":
        out_headers["Authorization"] = f"Bearer {secret}"
    elif attach.type == "query":
        path = _append_query(path, attach.name, secret)
    return out_headers, path


def _append_query(path: str, key: str, value: str) -> str:
    """Append ``key=value`` to ``path``'s querystring."""
    parts = urllib.parse.urlsplit(path)
    query = urllib.parse.parse_qsl(parts.query, keep_blank_values=True)
    query.append((key, value))
    new_query = urllib.parse.urlencode(query)
    return urllib.parse.urlunsplit(("", "", parts.path, new_query, parts.fragment))


def _do_upstream(
    *,
    policy: Policy,
    method: str,
    path: str,
    headers: dict[str, str],
    body: bytes,
    ssl_context: ssl.SSLContext | None,
) -> Response:
    """Issue the upstream HTTPS request synchronously and return a ``Response``."""
    conn = http.client.HTTPSConnection(
        policy.upstream.host,
        port=policy.upstream.port,
        timeout=policy.constraints.timeout_seconds,
        context=ssl_context,
    )
    try:
        conn.request(method, path, body=body, headers=headers)
        resp = conn.getresponse()
        body_out = resp.read()
        # http.client gives us a flat list of (name, value) pairs; group by name
        # so the wire response can carry multi-valued headers faithfully.
        grouped: dict[str, list[str]] = {}
        for name, value in resp.getheaders():
            grouped.setdefault(name, []).append(value)
        return Response(status=resp.status, headers=grouped, body=body_out)
    finally:
        conn.close()


def _err(code: str, message: str) -> dict[str, str]:
    """Build an error envelope for the wire."""
    return {"error": code, "message": message}
